### ISSUE TITLE
Reduce number of required methods in LanguageServer trait

### ISSUE THREAD

**Issue by icsaszar** [ebkalderon/tower-lsp#100]
I think that as the number of implemented methods grows it will be a lot of boilerplate to write for the users of the library to implement a simple server.

The simplest solution would probably be to require only the methods are declared (as a capability) when `ServerCapabilities::default()` is used.

Another approach could be to split up `LanguageServer` in the categories outlined in the specification like `WindowHandler`, `TelemetryHandler`, `WorkspaceHandler`, etc. leaving only `initialize` and `initialized` in `LanguageServer`. These (the new traits) shouldn't require implementing all methods either.

I think this is a great project and of all the other options I've considered for a language server it  has the simplest and most ergonomic API.

**Comment by icsaszar** [comment]
Another nice thing would be to have the default implementations to show an informative error message like "perhaps the capability was set to true, but the method is not implemented".

**Comment by ebkalderon** [comment]
Nice suggestions all around! I'm personally leaning towards [the last option](https://github.com/ebkalderon/tower-lsp/issues/100#issuecomment-589624719) the most. Simply changing all of the trait methods to have a default implementation, apart from perhaps `initialize()`, that defaults to an informative error message would provide the greatest value in usability for no additional overhead.

In regards to the first suggestion, I believe that it may be possible to whitelist server methods that have been declared as a capability in `ServerCapabilities` on startup, used in conjunction with providing default implementations for all other methods, but detecting a missing server method would still only occur at run-time so the overall safety improvement for this would be marginal at best, at the cost of increased library complexity.

I had actually considered splitting up the `LanguageServer` into multiple traits when I drafted the design for this library, but I determined that this approach would be too complex. The core problem is that `LspService` needs to be generic over some interface that represents a subset of the LSP protocol, and it is unclear how this would work if we had multiple potentially optional traits to support. It would require us to have some kind of base trait, e.g. `LanguageServer` and all other traits, e.g. `WindowHandler`, `TelemetryHandler`, `WorkspaceHandler`, would have to inherit from this `LanguageServer` trait. `LspService` would then remain generic over `LanguageServer`, but would need some mechanism to dispatch requests from the other traits. One could achieve this by adding another method to `LanguageServer` which the other traits' methods could use to emit responses, but this would leak library implementation details to the user.

**Comment by icsaszar** [comment]
> In regards to the first suggestion, I believe that it may be possible to whitelist server methods that have been declared as a capability in ServerCapabilities on startup, used in conjunction with providing default implementations for all other methods, but detecting a missing server method would still only occur at run-time so the overall safety improvement for this would be marginal at best, at the cost of increased library complexity.

I was thinking about this too, but since a method that is not declared as a capability shouldn't be called by clients anyway, the implementation I can think of is going through all the capabilities (which is super tedious because each capability needs to be matched to the corresponding method by hand) and calling each method that is declared to see if it's implemented. The problem would then be the parameters/values to call the methods with to not break the library client's implementation on startup.

**Comment by ebkalderon** [comment]
Agreed. The performance impact at initialization time could be non-trivial, and I would have to maintain hand-crafted mappings of all available LSP capabilities to trait methods in this library and review for correctness every time `lsp-types` releases a new version. I don't think that is particularly manageable nor maintainable long-term. 😄 

Ultimately, I think the last option you had suggested solves the problem in the issue description adequately and is the simplest to implement out of the three by a large margin.

**Comment by icsaszar** [comment]
If you have an error message in mind I can get started with implement this.

**Comment by ebkalderon** [comment]
I think it would be best to return a `jsonrpc_core::Error` saying that the given method is not implemented on the server and also log a message with `error!()` that an unimplemented method was called by the client. Panicking upon receiving an unimplemented method would mean that any misbehaving LSP client would be capable of causing a DoS on the server.

**Comment by ebkalderon** [comment]
@icsaszar Pull request #101 has landed in `master`, and `tower-lsp` has just been converted to async/await! This means we should be able to continue with resolving this issue.

As stated in the previous comment, I believe that default method implementations should be provided for all request methods, except for `initialize()`. The default implementation should return a `jsonrpc_core::Error::method_not_found()`, which corresponds to error code `-32601` as per the [JSON-RPC spec](https://www.jsonrpc.org/specification#error_object). We should also log a message with `error!()` that an unimplemented server method was called by the client.

Does this sound reasonable to you?

**Comment by ebkalderon** [comment]
However, I think we should treat the `shutdown()` method differently and have it return an `Ok(())` by default, since that method is not gated by a server capability, and I imagine that it would be surprising for a simple `tower-lsp` server to return an error when `shutdown()` is called.

**Comment by icsaszar** [comment]
I just submitted #103 with an initial implementation. I think it is fine to require shutdown, at least to remind people that they should do some cleanup :).

**Comment by ebkalderon** [comment]
Sure, I guess that's a good enough tradeoff! 😄

**Comment by ebkalderon** [comment]
Closed by #103.

**Comment by ebkalderon** [comment]
@icsaszar @Marwes For context regarding my earlier thoughts about breaking down the `LanguageServer` trait into granular pieces, I had considered providing multiple traits with their own `initialize()` methods, which would accept only the relevant parts of the `InitializeParams` as an argument and return only the relevant parts of `InitializeResult`. The `LspService`, then, would cobble together the appropriate response to give to the client.

If an implementation of a particular trait is omitted, the `LspService` will specify an appropriate default value to `InitializeResult` disabling the feature(s) in question.

See the hypothetical usage below:

```rust
#[derive(Default)]
struct Backend {
    workspace: Mutex<codespan::Files>,
    some_other_state: RwLock<SomeOtherState>,
}

impl LanguageServer for Backend {
    fn initialize(&self, params: InitializeParamsWithoutClientCaps) -> Result<InitializeResultWithoutClientCaps> {
        Ok(...)
    }

    async fn shutdown(&self) -> Result<()> {
        Ok(())
    }
}

#[derive(Default)]
struct BackendSync;

impl TextDocumentSync<Backend> for BackendSync {
    fn initialize(&self, state: &Backend, params: Option<SynchronizationCapability>) -> Result<Option<TextDocumentSyncCapability>> {
        Ok(TextDocumentSyncCapability::Kind(TextDocumentSyncKind::Full))
    }
}

#[derive(Default)]
struct BackendExecuteCommand;

impl ExecuteCommand<Backend> for BackendExecuteCommand {
    // `params` corresponds to the client capabilities for executing commands
    // Return value corresponds to `InitializeResult::capabilities::execute_command_provider`
    fn initialize(&self, state: &Backend, params: Option<GenericCapability>) -> Result<ExecuteCommandOptions> {
        Ok(...)
    }

    async fn execute_command(&self, state: &Backend, client: &Client, params: ExecuteCommandParams) -> Result<Option<Value>> {
        let _files = state.files.lock().await;
        Ok(...)
    }
}

// `Backend` implements `LanguageServer`
let (service, messages) = LspService::new(Backend::default())
    .text_document_sync(BackendSync::default())
    .execute_command(BackendExecuteCommand::default())
    .build();
```

Depending on whether implementations are provided in the `LspService` construction above, the server will cobble together a semantically correct `InitializeResult` value at runtime and respond to the client appropriately. If a trait implementation for a particular feature or set of features isn't provided to the builder, the `LspService` would set the appropriate provider value in the `InitializeResult` returned to the client to disable it, e.g. `None` or `Some(false)`.

I decided early on not to pursue this approach for `tower-lsp`, as it added quite a bit of internal complexity and felt rather awkward to use. But I thought maybe you might be interested in reading about it nonetheless, in case it sparks any further ideas.
